### PR TITLE
Vulkan implementation (via Kompute)

### PR DESCRIPTION
**Pull request by niansa**
Hey!

This is an attempt for a Vulkan implementation in GLSL via Kompute. The use of Kompute instead of using Vulkan directly is to avoid thousands of lines of hard to maintain boilterplate code.

I mostly base the code on the current Metal implementation and approach.

## Already implemented
 - `ADD` (Untested)
 - `MUL` (Untested)
 - `SCALE` (Test passed) ✅
 - `SILU` (Test passed) ✅
 - `RELU` (Test passed) ✅
 - `GELU` (Test passed) ✅
 - `SOFT_MAX` (Test failed) :x:
 - `DIAG_MASK_INF` (likely broken)
 - `MUL_MAT` (still needs fallback for small sizes)
   - `F16` (Unused)
   - `Q4_0` (Test failed) :x:
   - `Q4_1` (Test passed) ✅
 - `GET_ROW`
   - `F16` (Unused)
   - `Q4_0` (Test passed) ✅
   - `Q4_1` (Test failed) :x:
 - `NORM` (Unused)
 - `RMS_NORM` (Test passed) ✅
 - `ROPE` (Test passed) ✅
 - `CPY` (Test passed) ✅

## TODO things before merge
 - Remove Kompute submodule (expect it from system!) 
 - Clean up 
 - Fix Vulkan validation errors
 - Fix conflicts

### CODE REVIEW

**Comment by 0cc4m** [comment]
Hey, very cool. I'm also working on a Vulkan implementation, although one using pure Vulkan, based on the current OpenCL backend. I wasn't even aware of Kompute, will check this out definitely.

File: ggml-vulkan.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+                       const char * name,
+                             void * data,
+                           size_t   size,
+                           size_t   max_size);
+
+void ggml_vk_h2d_tensor(struct ggml_kompute_context * ctx, struct ggml_tensor * t);
+void ggml_vk_d2h_tensor(struct ggml_kompute_context * ctx, struct ggml_tensor * t);
+
+void ggml_vk_dequantize_row_q4_0(const void * x, float * y, int k);
+void ggml_vk_dequantize_row_q4_1(const void * x, float * y, int k);
```
Comment (ggerganov) [inline ggml-vulkan.h:36]: These should not be needed

File: ggml.c [hunk 2]
```diff
@@ -161,6 +161,8 @@ inline static void* ggml_aligned_malloc(size_t size) {
 #endif
 #elif defined(GGML_USE_OPENBLAS)
 #include <cblas.h>
+#elif defined(GGML_USE_KOMPUTE)
+#include "ggml-vulkan.h"
```
Comment (ggerganov) [inline ggml.c:165]: Ideally, you don't want to include `ggml-vulkan.h` in `ggml.c` similar to the Metal implementation

File: ggml.c [hunk 3]
```diff
@@ -1548,15 +1550,15 @@ static void ggml_vec_dot_q8_0_q8_0(const int n, float * restrict s, const void *
...
+        .dequantize_row_q         = (dequantize_row_q_t) ggml_vk_dequantize_row_q4_0,
         .quantize_row_q           = quantize_row_q4_0,
         .quantize_row_q_reference = (quantize_row_q_t) quantize_row_q4_0_reference,
         .quantize_row_q_dot       = quantize_row_q8_0,
         .vec_dot_q                = ggml_vec_dot_q4_0_q8_0,
         .vec_dot_type             = GGML_TYPE_Q8_0,
     },
     [GGML_TYPE_Q4_1] = {
-        .dequantize_row_q         = (dequantize_row_q_t)dequantize_row_q4_1,
+        .dequantize_row_q         = (dequantize_row_q_t) ggml_vk_dequantize_row_q4_1,
```
Comment (ggerganov) [inline ggml.c:1561]: Should not change these - I assume these are temporary

File: llama.cpp [hunk 4]
```diff
@@ -806,6 +810,14 @@ struct llama_model_loader {
...
                     }
                     break;
+#elif defined(GGML_USE_KOMPUTE)
+                case GGML_BACKEND_GPU:
+                    lt.ggml_tensor->data = lt.data;
+                    ggml_vk_h2d_tensor(lctx.ctx_kompute, lt.ggml_tensor);
+                    if (!use_mmap) {
+                        free(lt.data);
+                    }
+                    break;
```
Comment (ggerganov) [inline llama.cpp:820]: I don't think you need this - you are already passing the "data" buffer later in the code

File: llama.cpp [hunk 5]
```diff
@@ -1311,7 +1323,7 @@ static void llama_model_load_internal(
     }
 #endif
 
-    ml->load_all_data(progress_callback, progress_callback_user_data, use_mlock ? &lctx.model.mlock_mmap : NULL);
+    ml->load_all_data(lctx, progress_callback, progress_callback_user_data, use_mlock ? &lctx.model.mlock_mmap : NULL);
```
Comment (ggerganov) [inline llama.cpp:1326]: Not needed

**Review by ggerganov — COMMENTED** [review]
Cool stuff!!

I think some of the changes in `llama.cpp` and `ggml.c` are not necessary - assuming this is just a work in progress. I might be missing some detail though - so let me know why you need them if that is the case

**Comment by ggerganov** [comment]
> Can I add padding fields to the block structs ifdef GGML_KOMPUTE to allow GLSL to read these structs? Would you accept a change like this or would it even break other stuff?

Should avoid this. I guess it is something related to alignment. I think we had a similar issue in the OpenCL implementation where we introduce padded structs (not sure if this is still the case). Before doing anything, we try to make it work with the existing structs

**Comment by 0cc4m** [comment]
You can find my WIP Vulkan branch [here](https://github.com/0cc4m/koboldcpp/tree/vulkan) , I approached it from the other side. I want feature-parity with OpenCL first, so I already have a matrix multiplication shader (although not done optimizing yet).

File: ggml.c (hunk 3 above)
Comment (niansa) [reply ggml.c:1561]: Yup

**Comment by niansa** [comment]
@ggerganov please note this repo is "not ready for review"
But thanks for the hints, I appreciate them :+1:

**Comment by niansa** [comment]
> You can find my WIP Vulkan branch [here](https://github.com/0cc4m/koboldcpp/tree/vulkan) , I approached it from the other side. I want feature-parity with OpenCL first, so I already have a matrix multiplication shader (although not done optimizing yet).

@0cc4m Are you interested in a bit of a knowledge exchange? I'm pretty much just a beginner to this whole topic, so I'd be really interested in this.
Feel free to text me on Discord `tuxifan`, Telegram `@tuxifan`, Libera Chat IRC `tuxifan`/`nisa` or E-Mail `tuxifan@posteo.de`. :-)

**Comment by 0cc4m** [comment]
@niansa Of course, I sent you a friend request on Discord.

**Comment by SlyEcho** [comment]
> > Can I add padding fields to the block structs ifdef GGML_KOMPUTE to allow GLSL to read these structs? Would you accept a change like this or would it even break other stuff?
>
> Should avoid this. I guess it is something related to alignment. I think we had a similar issue in the OpenCL implementation where we introduce padded structs (not sure if this is still the case). 

True, we had a similar issue in OpenCL, the solution was to copy the data into a new array which added padding. Obviously not ideal, but if the weight is loaded into the device it may not be that significant.

But in the end the issue was solved by adding an attribute to pack the struct which the OpenCL language actually supports by spec.

**Comment by AlphaAtlas** [comment]
👀 

Other than reaching more platforms, what advantages would this Vulkan implementation have over OpenCL? Better performance? Better support for IGPs? Support for the matrix acceleration in Arc/RDNA3?

**Comment by SlyEcho** [comment]
OpenCL and OpenGL are basically deprecated and Vulkan is the replacement. Better driver support in newer GPUs etc.

**Comment by ggerganov** [comment]
> Other than reaching more platforms, what advantages would this Vulkan implementation have over OpenCL? Better performance? Better support for IGPs? Support for the matmul acceleration in Arc/RDNA3?

In addition to what @SlyEcho answered, my view is that since we now have a relatively good way to implement decoupled GPU backends (CUDA, Metal, OpenCL, etc.), there is no good reason not to keep doing it. As long as the new implementations do not touch the core `ggml` API and don't break the general workflow of other backends, we can keep adding support for new backends. Developers can join and help maintain the respective implementations. We can even have multiple implementations for a given single backend since we are still learning and finding new ways to make things more optimal and we don't know yet what is "the best way".

In the long run, we can eventually obsolete and deprecate certain implementations, but for now the main goal is to experiment and explore the most efficient ways for hardware acceleration. Supporting more backends also helps to find certain good patterns in the implementations and even though currently there is a lot of "copy-paste", at some point we can think about the best way to consolidate the code and reuse the best techniques that we have found across all architectures.

**Comment by Firstbober** [comment]
I can't get both this PR and #2059 to work. I wanted to run a perplexity test with `wiki.test.raw` as recommended in README, but all I get is GGML_ASSERT on `ggml-vulkan.cpp:142: res != ctx->tensors.end()`. Furthermore, I also tested just normal inference, but it just gets out of memory on my ram (16 GiB physical + 16 GiB swap).

In the case of second implementation, I can't even get to filling my ram as it's just crashes on me with `ggml_vulkan: vk_transfer_queue_family_index invalid` and i can't do anything.

I am using RX 580 (RADV driver, Polaris 10 architecture). On the side note, OpenCL works, but prompt ingestion is painfully slow.

**Comment by 0cc4m** [comment]
@Firstbober AMD GPUs have fewer queues, it seems. The initialization code of my implementation didn't work with that yet. Can you try again? Don't expect any good performance yet, though.

**Comment by Firstbober** [comment]
> @Firstbober AMD GPUs have fewer queues, it seems. The initialization code of my implementation didn't work with that yet. Can you try again? Don't expect any good performance yet, though.

It did work this time after your patch! I needed to remove the validation layer as the amount of errors it produced was too much. Two major ones were `VUID-vkCmdDispatch-groupCountX-00386(ERROR / SPEC)` and `VUID-VkDeviceCreateInfo-queueFamilyIndex-02802(ERROR / SPEC)`.

After making the output readable, I can clearly see my GPU working and ingesting the prompt. I wasn't able to test the perplexity of `wiki.test.raw` as it just OOMed me (`vk::Device::allocateCommandBuffers: ErrorOutOfHostMemory`), but I only have 4 GiB of VRAM, so it's pretty understandable.

It seems to be very close to OpenCL. I ran the test with `./main -n 1` (for OpenCL, I also added `-ngl 40`):
Vulkan:
```
llama_print_timings:        load time =  2129,57 ms
llama_print_timings:      sample time =     1,13 ms /     1 runs   (    1,13 ms per token,   882,61 tokens per second)
llama_print_timings: prompt eval time = 45162,26 ms /   372 tokens (  121,40 ms per token,     8,24 tokens per second)
llama_print_timings:        eval time =     0,00 ms /     1 runs   (    0,00 ms per token,      inf tokens per second)
llama_print_timings:       total time = 45182,14 ms
```

OpenCL:
```
llama_print_timings:        load time =  9512,41 ms
llama_print_timings:      sample time =     1,02 ms /     1 runs   (    1,02 ms per token,   985,22 tokens per second)
llama_print_timings: prompt eval time = 47387,69 ms /   372 tokens (  127,39 ms per token,     7,85 tokens per second)
llama_print_timings:        eval time =     0,00 ms /     1 runs   (    0,00 ms per token,      inf tokens per second)
llama_print_timings:       total time = 47411,72 ms
```

Overall, Vulkan used my VRAM much better, while OpenCL allocated all of it and even didn't use my GPU as much as your Vulkan implementation.

**Comment by JianbangZ** [comment]
How do I test this branch out?
When I do CMake, it complains Could NOT find Vulkan (missing: Vulkan_LIBRARY Vulkan_INCLUDE_DIR) (found                                               version "")

**Comment by JianbangZ** [comment]
I couldn't get this to work, which throws some werid error "fatal error C1189: #error:  You
r C implementation is not IEC 559 compliant, which is required for proper Vulkan interop"

I also came around another project which fully uses Vulkan, and I tested on Windows with Intel integrated GPU. Speed is quite plausible but the good thing is it fully used the integrated GPU so PC doesn't feel slow anymore, otherwise CPU will be at 100% https://mlc.ai/mlc-llm/

**Comment by AlphaAtlas** [comment]
@JianbangZ 

Yeah, MLC uses Apache TVM Vulkan as a backend. It features are pretty barebones, and it doesn't support any model splitting, CPU offloading, or anything like K Quants, but it is *very* fast and relatively portable.

If you are using MSVC or something, my guess is you need to compile in WSL?

**Comment by JianbangZ** [comment]
@AlphaAtlas I use https://github.com/skeeto/w64devkit

**Comment by AlphaAtlas** [comment]
> I use https://github.com/skeeto/w64devkit

@JianbangZ 🤷 

You could try reporting this issue to w64devkit, but it might be a fundamental compiler limitation.

**Comment by SlyEcho** [comment]
> I wasn't able to compile Kompute due to compile warnings, so it was necessary to edit kompute/CMakeLists.txt (after updating the kompute submodule) and remove -Werror from the build flags.

The error/warning seems to come from the fmt library build, using the system's fmt library with `-KOMPUTE_OPT_USE_BUILT_IN_FMT=OFF` allowed me to build. Also useful: `-DKOMPUTE_OPT_LOG_LEVEL=Off`.

**Comment by riverzhou** [comment]
[Jul 12 2023 15:12:23] [warn] [/home/river/LLM/llama.cpp/kompute/src/Manager.cpp:231] Kompute Manager no valid layer names found from desired layer names
[Jul 12 2023 15:12:23] [info] [/home/river/LLM/llama.cpp/kompute/src/Manager.cpp:339] Using physical device index 0 found llvmpipe (LLVM 15.0.7, 256 bits)

What's problem about this?

**Pull request closed without merging**